### PR TITLE
Acohen/multi blade row

### PR DESCRIPTION
**Pull request by Acohen-work**
1. Rename MBR batch scripts to reflect the batch nature of the processing.
2. Introduce MBR control module and tests
3. Update packaging and build tools to use poetry

### CODE REVIEW

File: examples/mbr_sweep_example.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+from ansys.turbogrid.core.multi_blade_row.multi_blade_row import multi_blade_row as MBR
+
+
+all_face_areas = {}
+all_element_counts = {}
+
+
+#################################################################################
+# Create a multi_blade_row and initialize it
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (swilson-work) [inline examples/mbr_sweep_example.py:54]: @Acohen-work Make sure you use the period at the end of the sentences consistently. Also make sure that the underlining is the same length as the heading text.

File: pyproject.toml [hunk 2]
```diff
@@ -1,77 +1,86 @@
...
-
-authors = [
-    {name = "ANSYS, Inc.", email = "pyansys.core@ansys.com"},
-]
-maintainers = [
-    {name = "ANSYS, Inc.", email = "pyansys.core@ansys.com"},
-]
-
+license = "MIT"
+authors = ["ANSYS, Inc. <ansys.support@ansys.com>"]
```
Comment (swilson-work) [inline pyproject.toml:11]: These changes to the authors/maintainers do not match the template.

File: tests/test_multi_blade_row.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+install_path = pathlib.PurePath(__file__).parent.parent.as_posix()
+
+
+def test_multi_blade_row_basic():
+    machine = MBR()
+    machine.init_from_ndf(f"{install_path}/tests/ndf/AxialFanMultiRow.ndf")
+    # This is a planned point-data initialization file format
+    # machine.init_from_tgmachine(f"{dir_path}/~.TGMachine")
+    # blade_rows = machine.get_blade_row_names()
+    print(f"Average Face Area Before: {machine.get_average_base_face_areas()}")
```
Comment (swilson-work) [inline tests/test_multi_blade_row.py:40]: All instances of "cannonical" should be replaced with "canonical".

File: examples/mbr_sweep_example.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+from ansys.turbogrid.core.multi_blade_row.multi_blade_row import multi_blade_row as MBR
+
+all_face_areas = {}
+all_element_counts = {}
+
+
+#################################################################################
+# Create a multi_blade_row and initialize it
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Use the Concepts NREC sample provided
```
Comment (swilson-work) [inline examples/mbr_sweep_example.py:54]: Make sure all sentences end with a period.

File: pyproject.toml [hunk 5]
```diff
@@ -1,87 +1,90 @@
...
-]
+[tool.poetry.dependencies]
+python = ">=3.9,<4.0"
+matplotlib = ">=3.5.1"
+importlib-metadata = ">=4.0"
+fabric = ">=3.2.2"
+Jinja2 = ">=3.1.2"
+ansys-turbogrid-api = ">=0.4.3"
+# ansys-turbogrid-api = { path = "../ansys-api-turbogrid/", develop = true } # Uncomment this if you have access to a local repo
+toml = "0.10.2" # This should be picked up from ansys-turbogrid-api one its lock file is updated
```
Comment (swilson-work) [inline pyproject.toml:38]: Typo "one" instead of "once"

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 6]
```diff
@@ -21,1083 +21,531 @@
...
+from ansys.turbogrid.core.launcher.launcher import launch_turbogrid, launch_turbogrid_container
+from ansys.turbogrid.core.multi_blade_row.single_blade_row import single_blade_row
+import ansys.turbogrid.core.ndf_parser.ndf_parser as ndf_parser
 
-    #: An alias for the TurboGridLocationType in pyturbogrid_core.PyTurboGrid.
-    TurboGridLocationType = pyturbogrid_core.PyTurboGrid.TurboGridLocationType
 
-    #: When working in Ansys Labs, the name with full path of the file containing the container key.
-    tg_container_key_file = ""
+class Helpers:
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:50]: @Acohen-work Many of these helpers are not directly connected to multi_blade_row. They should be put into their own class/file for clarity and re-use.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 7]
```diff
@@ -21,1083 +21,531 @@
...
+            futures = [
+                executor.submit(self.__quit__, val) for key, val in self.tg_worker_instances.items()
+            ]
+            concurrent.futures.wait(futures)
 
-            Here each blade row has to be identified by the name of the main blade in the row.
-        """
-        self._custom_blade_settings = custom_blade_settings
+    # a A TGInit parser would need to be written for this to work
+    # def init_from_tginit(self, tginit_path: str):
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:128]: These commented lines should be removed.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 8]
```diff
@@ -21,1083 +21,531 @@
...
-            { blade_name : [("Full CCL Object Path", "Param Name=Value"), ... ], ... }
+    @staticmethod
+    def transfer_files_from_container(
+        container_connection: Connection, local_folder_path: str, remote_filenames: list
+    ):
+        for filename in remote_filenames:
+            Helpers.transfer_file_from_container(container_connection, filename, local_folder_path)
+
+
+class multi_blade_row:
```
Comment (swilson-work) [inline src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:97]: @Acohen-work There don't appear to be many doc strings in this class/file.

File: src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py [hunk 9]
```diff
@@ -21,1083 +21,531 @@
...
+        self.ndf_file_name, self.ndf_file_extension = os.path.splitext(ndf_name)
+        print(f"{ndf_name=}")
+
+        if use_existing_tginit_cad == False:
+            tg_port = None
+            tg_execution_control = None
+            if (
+                self.turbogrid_location_type
+                == PyTurboGrid.TurboGridLocationType.TURBOGRID_RUNNING_CONTAINER
+            ):
```
[the remark this answers is not in the document]
Comment (Acohen-work) [reply src/ansys/turbogrid/core/multi_blade_row/multi_blade_row.py:149]: The helpers are moved. Once we have gRPC we can use the ansys product instance launcher to simplify some of this as well, or write our own launch/file management system.

**Review by swilson-work — APPROVED** [review]

**Pull request merged**